### PR TITLE
[apex] New Rule: Queueable Should Attach Finalizer

### LINKED ISSUE #5302: [apex] New Rule: Queueable Should Attach Finalizer
**Proposed Rule Name:** Queueable Should Attach Finalizer

**Proposed Category:** Best Practices

**Description:** We want to warn the author of Apex whenever they write a class implementing the `Queueable` interface and they fail to attach a `Finalizer`. We should explain that the lack of a `Finalizer` provides no handling capabilities should the `Queueable` action fail.

**Code Sample:** 
```java
// Improper code, does not call `System.attachFinalizer` within the `execute` method.
public class UserUpdater implements Queueable {
  private List<User> usersToUpdate;

  public UserUpdater(List<User> usersToUpdate) {
    this.usersToUpdate = usersToUpdate;
  }

  public void execute(QueueableContext context) {
    update usersToUpdate;
  }
}

// Properly attaches a `Finalizer`
public class UserUpdater implements Queueable, Finalizer {
  private List<User> usersToUpdate;

  public UserUpdater(List<User> usersToUpdate) {
    this.usersToUpdate = usersToUpdate;
  }

  public void execute(QueueableContext context) {
    System.attachFinalizer(this);
    update usersToUpdate;
  }

  public void execute(FinalizerContext ctx) {
    if (ctx.getResult() == ParentJobResult.SUCCESS) {
      // Handle success
    } else {
      // Handle failure
    }
  }
}


```
**Possible Properties:**

*   Should this rule be customizable via properties? - No, this is a straightforward rule with no properties.

### PR DESCRIPTION
**Pull request by mitchspano**
## Describe the PR

Warns the author of Apex whenever they write a class which implements the `Queueable` interface and do not attach a `Finalizer`.

## Related issues

- Fixes #5302

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-apex/src/main/resources/category/apex/bestpractices.xml [hunk 1]
```diff
@@ -285,4 +285,31 @@ Detects when a local variable is declared and/or assigned but not used.
...
     </rule>
 
+    <rule name="QueueableShouldAttachFinalizer"
+          since="7.8.0"
+          language="apex"
+          message="It is best practice to call the `System.attachFinalizer(Finalizer f)` method within the `execute` method of a class which implements the `Queueable` interface."
+          class="net.sourceforge.pmd.lang.apex.rule.bestpractices.QueueableShouldAttachFinalizerRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_apex_bestpractices.html#queueableshouldattachfinalizer">
+        <description>
+Detects when the Queueable interface is used but a Finalizer is not attached.
```
Comment (adangel) [inline pmd-apex/src/main/resources/category/apex/bestpractices.xml:295]:
```suggestion
Detects when the Queueable interface is used but a Finalizer is not attached.

It is best practice to call the `System.attachFinalizer(Finalizer f)` method within the `execute` method of a class which implements the `Queueable` interface.
Without attaching a Finalizer, there is no way of designing error recovery actions should the Queueable action fail.
```

File: pmd-apex/src/main/resources/category/apex/bestpractices.xml [hunk 2]
```diff
@@ -285,4 +285,31 @@ Detects when a local variable is declared and/or assigned but not used.
         </example>
     </rule>
 
+    <rule name="QueueableShouldAttachFinalizer"
```
Comment (adangel) [inline pmd-apex/src/main/resources/category/apex/bestpractices.xml:288]: Maybe "QueueableWithouFinalizer" as rule name? That sounds more neutral and just expresses, what the rule detects.

```suggestion
    <rule name="QueueableWithoutFinalizer"
```

File: pmd-apex/src/main/resources/category/apex/bestpractices.xml [hunk 3]
```diff
@@ -285,4 +285,31 @@ Detects when a local variable is declared and/or assigned but not used.
...
+          since="7.8.0"
+          language="apex"
+          message="It is best practice to call the `System.attachFinalizer(Finalizer f)` method within the `execute` method of a class which implements the `Queueable` interface."
```
Comment (adangel) [inline pmd-apex/src/main/resources/category/apex/bestpractices.xml:291]:
```suggestion
          message="This Queueable doesn't attach a Finalizer"
```

File: pmd-apex/src/main/resources/category/apex/bestpractices.xml [hunk 4]
```diff
@@ -285,4 +285,31 @@ Detects when a local variable is declared and/or assigned but not used.
...
+        public UserUpdater(List<User> usersToUpdate) {
+            this.usersToUpdate = usersToUpdate;
+        }
+
+        public void execute(QueueableContext context) { // no Finalizer is attached
+            update usersToUpdate;
+        }
+    }
+]]>
+        </example>
```
Comment (adangel) [inline pmd-apex/src/main/resources/category/apex/bestpractices.xml:312]: Maybe we should add a second example, that shows, how to do it correctly?

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableShouldAttachFinalizerRule.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  private static final String SYSTEM_ATTACH_FINALIZER = "system.attachfinalizer";
+
+  /** Scans the top level class and all inner classes. */
+  @Override
+  public Object visit(ASTUserClass topLevelClass, Object data) {
+    scanClassForViolation(topLevelClass, data);
+    for (ASTUserClass innerClass : topLevelClass.descendants(ASTUserClass.class).toList()) {
+      scanClassForViolation(innerClass, data);
+    }
+    return data;
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableShouldAttachFinalizerRule.java:35]: This sounds like a [rule-chain](https://docs.pmd-code.org/latest/pmd_userdocs_extending_writing_java_rules.html#economic-traversal-the-rulechain) rule implementation. Therefore, it should register itself via 

```java
    @Override
    protected @NonNull RuleTargetSelector buildTargetSelector() {
        return RuleTargetSelector.forTypes(ASTUserClass.class);
    }
```

Then you can get rid of the manual for-loop (as the visit method will then be called for each ASTUserClass anyway).

**Review by adangel — CHANGES_REQUESTED** [review]
Thanks for the PR!

I think, the rule documentation should be improved a bit. I'm suggesting a slightly different (more neutral) name. We have also some more guidelines documented at https://docs.pmd-code.org/latest/pmd_devdocs_major_rule_guidelines.html

The goal is, that the rules are easily understandable, what they detect, why they detect it, and what the alternative ("correct") code would be...

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableShouldAttachFinalizerRule.java (hunk 5 above)
Comment (mitchspano) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableShouldAttachFinalizerRule.java]: Great call, thanks for pointing this out! I've updated the implementation to use the rule chain.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableWithoutFinalizerRule.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        asCtx(data).addViolation(theMethod);
+      }
+    }
+    return data;
+  }
+
+  /** Determines if the class implements the Queueable interface. */
+  private boolean implementsTheQueueableInterface(ASTUserClass theClass) {
+    for (String interfaceName : theClass.getInterfaceNames()) {
+      if (interfaceName.equalsIgnoreCase(QUEUEABLE)) {
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableWithoutFinalizerRule.java:56]: PMD Violation: [LiteralsFirstInComparisons](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#literalsfirstincomparisons) -> Position literals first in String comparisons.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableWithoutFinalizerRule.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    return false;
+  }
+
+  /**
+   * Determines if the method is the `execute(QueueableContext context)` method. Parameter count is
+   * checked to account for method overloading.
+   */
+  private boolean isTheExecuteMethodOfTheQueueableInterface(ASTMethod theMethod) {
+    if (!theMethod.getCanonicalName().equalsIgnoreCase(EXECUTE)) {
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableWithoutFinalizerRule.java:68]: PMD Violation: [LiteralsFirstInComparisons](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#literalsfirstincomparisons) -> Position literals first in String comparisons.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableWithoutFinalizerRule.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+      return false;
+    }
+    List<ASTParameter> parameters = theMethod.descendants(ASTParameter.class).toList();
+    return parameters.size() == 1
+        && parameters.get(0).getType().equalsIgnoreCase(QUEUEABLE_CONTEXT);
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableWithoutFinalizerRule.java:73]: PMD Violation: [LiteralsFirstInComparisons](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#literalsfirstincomparisons) -> Position literals first in String comparisons.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableWithoutFinalizerRule.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  /** Determines if the method calls the `System.attachFinalizer(Finalizer f)` method. */
+  private boolean callsTheSystemAttachFinalizerMethod(ASTMethod theMethod) {
+    for (ASTMethodCallExpression methodCallExpression :
+        theMethod.descendants(ASTMethodCallExpression.class).toList()) {
+      if (methodCallExpression.getFullMethodName().equalsIgnoreCase(SYSTEM_ATTACH_FINALIZER)) {
```
Comment (adangel) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableWithoutFinalizerRule.java:80]: PMD Violation: [LiteralsFirstInComparisons](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#literalsfirstincomparisons) -> Position literals first in String comparisons.

**Review by adangel — CHANGES_REQUESTED** [review]
Thanks for the update!

However, this is not ready yet: There are some PMD violations, you need to fix first. And then, we'll probably see checkstyle errors (we use 4 space indentation, etc. - just run `./mvnw verify` before commit.... see also https://github.com/pmd/pmd/wiki/Setup-IDE)

**Comment by mitchspano** [comment]
@all-contributors please add @mitchspano for code

**Comment by mitchspano** [comment]
@adangel Thank you for your guidance here. I have resolved all PMD and checkstyle issues. `./mvnw verify` produces a passing result.

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to main:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 11 errors and 7 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/11864248813?pr=5303">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by adangel — APPROVED** [review]
Thanks, looks good!

**Pull request merged**